Loop/Managers/RemoteDataServicesManager.swift

```diff
@@ -6,165 +6,362 @@
+        clearPumpEventQueryAnchor(for: remoteDataService)
+        clearSettingsQueryAnchor(for: remoteDataService)
+        clearStatusQueryAnchor(for: remoteDataService)
     }
 
-    private func uploadSettings() {
-        guard !remoteDataServices.isEmpty else {
-            return
+}
+
+extension RemoteDataServicesManager {
+
+    private var carbDataType: String { return "Carb" }
+
+    public func carbStoreHasUpdatedCarbData(_ carbStore: CarbStore) {
+        remoteDataServices.forEach { self.synchronizeCarbData(from: carbStore, to: $0) }
+    }
+
+    private func synchronizeCarbData(from carbStore: CarbStore, to remoteDataService: RemoteDataService) {
+        dispatchQueue(for: remoteDataService, withDataType: carbDataType).async {
+            let semaphore = DispatchSemaphore(value: 0)
+            let queryAnchor = UserDefaults.appGroup?.getQueryAnchor(for: remoteDataService, withDataType: self.carbDataType) ?? CarbStore.QueryAnchor()
+
+            carbStore.executeCarbQuery(fromQueryAnchor: queryAnchor, limit: remoteDataService.carbDataLimit) { result in
+                switch result {
+                case .failure(let error):
+                    self.log.error("Error querying carb data: %{public}@", String(describing: error))
+                    semaphore.signal()
+                case .success(let queryAnchor, let deleted, let stored):
+                    remoteDataService.synchronizeCarbData(deleted: deleted, stored: stored) { result in
+                        switch result {
+                        case .failure(let error):
+                            self.log.error("Error synchronizing carb data: %{public}@", String(describing: error))
+                        case .success:
+                            UserDefaults.appGroup?.setQueryAnchor(for: remoteDataService, withDataType: self.carbDataType, queryAnchor)
+                        }
+                        semaphore.signal()
+                    }
+                }
+            }
+
+            semaphore.wait()
         }
+    }
 
-        guard let settings = UserDefaults.appGroup?.loopSettings else {
-            log.default("Not uploading due to incomplete configuration")
-            return
+    private func clearCarbQueryAnchor(for remoteDataService: RemoteDataService) {
+        dispatchQueue(for: remoteDataService, withDataType: carbDataType).async {
+            UserDefaults.appGroup?.deleteQueryAnchor(for: remoteDataService, withDataType: self.carbDataType)
         }
+    }
+
+}
+
+extension RemoteDataServicesManager {
+
+    private var doseDataType: String { return "Dose" }
+
+    public func doseStoreHasUpdatedDoseData(_ doseStore: DoseStore) {
+        remoteDataServices.forEach { self.synchronizeDoseData(from: doseStore, to: $0) }
+    }
 
-        remoteDataServices.forEach { $0.uploadSettings(settings, lastUpdated: lastSettingsUpdate) }
+    private func synchronizeDoseData(from doseStore: DoseStore, to remoteDataService: RemoteDataService) {
```

Comment:
This seems more of an upload (one way) than a synchronize (which implies two way).

Reply:
I will rename `upload...Data`.

Reply:
Done.